lib/scrapers/gmailscrape.py

```diff
@@ -0,0 +1,139 @@
+from tornado import gen
+from tornado.httpclient import HTTPError
+
+import httplib2
+import base64
+import email
+from lib.config import CONFIG
+
+from apiclient.discovery import build
+from oauth2client import client
+
+class GMailScraper(object):
+    name = 'gtext'
+
+    def __init__(self):
+        self.tokens = []
+        with open('lib/scrapers/snippets.txt', 'r') as fd:
+            self.tokens = [s.strip() for s in fd]
+
+    def get_content(raw):
+        data = base64.urlsafe_b64decode(raw)
+        email_parser = EmailParser(policy=policy.default)
+        email = email_parser.parsebytes(data)
+        plain = email.get_body(preferencelist=('plain',))
+        body = None
+        if plain:
+            body = plain.get_payload()
+        email_dict = dict(email)
+        email_dict['body'] = body
+        return email_dict
+
+
+    def paginate_messages(self, service, response):
+        threads = set()
+        if 'messages' in response:
+            for i in response['messages']:
+                threads.add(i['threadId'])
+
+        while 'nextPageToken' in response:
+            page_token = response['nextPageToken']
+            response = service.users().messages().list(userId='me',
+                    pageToken=page_token).execute()
+            for i in response['messages']:
+                threads.add(i['threadId'])
+        return messages
+
+    def get_recipient(self, email_data, parsed=False):
+        names = []
+        if parsed == True:
+            to_field = email_data['To']
+        else:
+            headers = email_data['payload']['headers']
+            to_field = [d['value'] for d in headers if d['name'] == 'To']
+        sent_to = to_field.split(', ')
+        for i in sent_to:
+            names.append(i[0].split(' <')[0])
+        return names
+
+    def get_raw_from_id(self, service, email_id):
+        try:
+            msg = service.users().messages().get(userId='me', id=email_id, format='raw').execute()
+            snip = msg['snippet']
+            email = self.get_content(msg['raw'])
+            #msg_str = base64.urlsafe_b64decode(msg['raw'].encode('ASCII'))
+            #mime_msg = email.message_from_string(msg_str)
+            #text = mime_msg.get_payload()[0]
+            return email, snip
+        except HTTPError as e:
```

Comment:
This exception handling doesn't actually help at all.... you catch the exception and then don't return anything so whatever called this function (which expects a tuple of `(email, snip)`) will the cause it's own exception and we're screwed.